File: oscar-cp/src/main/java/oscar/cp/constraints/Square.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints;

import oscar.algo.Inconsistency;
import oscar.cp.core.CPPropagStrength;
import oscar.cp.core.variables.CPIntVar;
import oscar.cp.core.Constraint;
import oscar.cp.core.variables.CPVar;
import scala.collection.Iterable;
import scala.jdk.javaapi.CollectionConverters;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class Square extends Constraint {

	private CPIntVar x;
	private CPIntVar y;

	/**
	 * x*x == y
	 * @param x
	 * @param y
	 * @see  CPIntVar#mul(cp.core.CPIntVar)
	 */
	public Square(CPIntVar x, CPIntVar y) {
		super(x.store(),"Square");
		this.x = x;
		this.y = y;
	}

	@Override
	public Iterable<CPVar> associatedVars() {
		List<CPVar> l = new LinkedList<>(Arrays.asList(x, y));
		return CollectionConverters.asScala(l);
	}

	@Override
	public void setup(CPPropagStrength l) throws Inconsistency {
		y.updateMin(0);
		propagate();
		if(isActive()) {
			if (!x.isBound()) {
				x.callPropagateWhenBoundsChange(this);
			}
			if (!y.isBound()) {
				y.callPropagateWhenBoundsChange(this);
			}
		}
	}

	@Override
	public void propagate() {
		// propagation of y
		int mx = x.getMin();
		int Mx = x.getMax();
		int mx2 = mx*mx;
		int Mx2 = Mx*Mx;

		//propagate y (which is not bound)
		if (mx >= 0) { // x will be positive
			y.updateMin(mx2);
			y.updateMax(Mx2);
		} else if (Mx <= 0) { // x is non positive
			y.updateMin(Mx2);
			y.updateMax(mx2);
		} else if (x.hasValue(0)) {
			//y min is already >= 0 (post does it)
			y.updateMax(Math.max(mx2, Mx2));
		} else {
			Integer a = (Integer) x.valueBefore(0);
			Integer b = (Integer) x.valueAfter(0);
			int a2 = a*a;
			int b2 = b*b;
			y.updateMin(Math.min(a2, b2));
			y.updateMax(Math.max(a2, b2));
		}
		//propagate x (which is not bound)
		int my = y.getMin();
		int My = y.getMax();
		int my2 = my*my;
		int My2 = My*My;

		int rootm = (int) (Mx <= 0 ? Math.ceil(Math.sqrt(my)) : Math.sqrt(my));
		int rootM = (int) Math.sqrt(My);

		if (mx >= 0) {
			x.updateMin(rootm);
			x.updateMax(rootM);
		} else if (Mx <= 0) {
			x.updateMax(-rootm);
			x.updateMin(-rootM);
		} else {
			x.updateMin(-rootM);

			x.updateMax(rootM);
			/*
			for (int v = -rootm+1; v < rootm; v++) {
				x.removeValue(v);
			}*/
		}
	}

}

File: src/main/scala/oscar/cp/constraints/Square.scala
package oscar.cp.constraints

import oscar.algo.Inconsistency
import oscar.cp.core.CPPropagStrength
import oscar.cp.core.Constraint
import oscar.cp.core.variables.CPIntVar
import oscar.cp.core.variables.CPVar

class Square(val x: CPIntVar, val y: CPIntVar) extends Constraint(x.store, "Square") {

  override def associatedVars(): Iterable[CPVar] = List(x, y)

  override def setup(l: CPPropagStrength): Unit = {
    y.updateMin(0)
    propagate()
    if (isActive) {
      if (!x.isBound) {
        x.callPropagateWhenBoundsChange(this)
      }
      if (!y.isBound) {
        y.callPropagateWhenBoundsChange(this)
      }
    }
  }

  override def propagate(): Unit = {
    val mx = x.getMin
    val Mx = x.getMax
    val mx2 = mx * mx
    val Mx2 = Mx * Mx

    if (mx >= 0) {
      y.updateMin(mx2)
      y.updateMax(Mx2)
    } else if (Mx <= 0) {
      y.updateMin(Mx2)
      y.updateMax(mx2)
    } else if (x.hasValue(0)) {
      y.updateMax(Math.max(mx2, Mx2))
    } else {
      val a = x.valueBefore(0)
      val b = x.valueAfter(0)
      val a2 = a * a
      val b2 = b * b
      y.updateMin(Math.min(a2, b2))
      y.updateMax(Math.max(a2, b2))
    }

    val my = y.getMin
    val My = y.getMax
    val my2 = my * my
    val My2 = My * My

    val rootm = if (Mx <= 0) Math.ceil(Math.sqrt(my)).toInt else Math.sqrt(my).toInt
    val rootM = Math.sqrt(My).toInt

    if (mx >= 0) {
      x.updateMin(rootm)
      x.updateMax(rootM)
    } else if (Mx <= 0) {
      x.updateMax(-rootm)
      x.updateMin(-rootM)
    } else {
      x.updateMin(-rootM)
      x.updateMax(rootM)
    }
  }
}
